File: problem2.cpp
#include<iostream>
#include<vector>
#include<queue>
#include<unordered_map>

using namespace std;

//bfs

class Solution {
    typedef struct gg{
        int x;
        int y;
        gg(int x,int y):x(x),y(y){}
    }gg;
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        vector<gg> dirs{
            gg(-1,0),
            gg(1,0),
            gg(0,1),
            gg(0,-1)
        };
        queue<gg> q;
        int m = mat.size();
        int n = mat.at(0).size();
        for(int i{};i<m;++i){
            for(int j{};j<n;++j){
                if(mat.at(i).at(j)==0) q.push(gg(i,j));
                if(mat.at(i).at(j) == 1) mat.at(i).at(j)=-1;
            }
        }
        int dist{1};
        while(!q.empty()){
            int sz = q.size();
            for(int i{};i<sz;++i){
                gg node = q.front();
                for(gg& dir: dirs){
                    int x = node.x +dir.x;
                    int y = node.y + dir.y;
                    if(x>=0 && x<m && y>=0 && y<n && mat.at(x).at(y) == -1){
                        mat.at(x).at(y) = dist;
                        q.push(gg(x,y));
                    }
                }
            q.pop();
            }
            ++dist;
        }
        return mat;
    }
};

// O(MN)^2 DFS

class Solution_2 {
    void dfs(vector<vector<int>>& mat,int nr, int nc,int m,int n,int val){
        if(nr>=0 && nr<n && nc>=0 && nc<m && mat.at(nr).at(nc)>val){
            mat.at(nr).at(nc) = val+1;
            dfs(mat,nr,nc+1,m,n,mat.at(nr).at(nc));
            dfs(mat,nr,nc-1,m,n,mat.at(nr).at(nc));
            dfs(mat,nr+1,nc,m,n,mat.at(nr).at(nc));
            dfs(mat,nr-1,nc,m,n,mat.at(nr).at(nc));
        }
    }
public:
    vector<vector<int>> updateMatrix(vector<vector<int>>& mat) {
        int n = mat.size();
        int m = mat.at(0).size();
        for(int i{};i<n;++i){
            for(int j{};j<m;++j){
                if(mat.at(i).at(j) == 1) mat.at(i).at(j) = m*n;
            }
        }
        for(int i{};i<n;++i){
            for(int j{};j<m;++j){
                if(mat.at(i).at(j)==0){
                    dfs(mat,i,j+1,m,n,0);
                    dfs(mat,i,j-1,m,n,0);
                    dfs(mat,i+1,j,m,n,0);
                    dfs(mat,i-1,j,m,n,0);
                }
            }
        }
        return mat;
    }
};

// optimal DFS

File: problem-2.go
/*Time Complexity : O(m * n)
Space Complexity : O(m * n)*/

func updateMatrix(mat [][]int) [][]int {
	// Check if the input matrix is empty
	if mat == nil || len(mat) == 0 || len(mat[0]) == 0 {
		// If so, return an empty matrix
		return [][]int{}
	}

	// Get the number of rows (m) and columns (n) in the matrix
	m, n := len(mat), len(mat[0])

	// Initialize an empty queue for BFS
	queue := make([][]int, 0)

	// Define the maximum possible value in the matrix
	MAX_VALUE := m * n

	// Iterate through each cell in the matrix
	for i := 0; i < m; i++ {
		for j := 0; j < n; j++ {
			// If the cell value is 0, add its coordinates to the queue
			if mat[i][j] == 0 {
				queue = append(queue, []int{i, j})
			} else {
				// Otherwise, set the cell value to MAX_VALUE
				mat[i][j] = MAX_VALUE
			}
		}
	}

	// Define possible movements: up, down, left, and right
	directions := [][]int{{1, 0}, {-1, 0}, {0, 1}, {0, -1}}

	// Perform BFS traversal
	for len(queue) > 0 {
		// Dequeue a cell from the queue
		cell := queue[0]
		queue = queue[1:]

		// Explore neighboring cells
		for _, dir := range directions {
			r, c := cell[0]+dir[0], cell[1]+dir[1]
			// Check if the neighboring cell is within bounds and if updating its value minimizes distance
			if r >= 0 && r < m && c >= 0 && c < n && mat[r][c] > mat[cell[0]][cell[1]]+1 {
				// Update the neighboring cell's value
				queue = append(queue, []int{r, c})
				mat[r][c] = mat[cell[0]][cell[1]] + 1
			}
		}
	}

	// Return the updated matrix
	return mat
}
